src/test/groovy/pl/allegro/tech/opel/OpelEngineMathIntegrationSpec.groovy

```diff
@@ -139,6 +139,36 @@ class OpelEngineMathIntegrationSpec extends Specification {
         "identity(if (false || (true && true)) true else false)" || true
     }
 
+    @Unroll
+    def '! should negate boolean expression #input'() {
+        given:
+        def engine = create()
+                .withImplicitConversion(String, BigDecimal, { string -> new BigDecimal(string) })
+                .withImplicitConversion(BigDecimal, String, { decimal -> decimal.toPlainString() })
+                .build()
+        def evalContext = EvalContextBuilder.create()
+                .withCompletedValue('value', true)
+                .withCompletedValue('fun', constFunctionReturning(false))
+                .build()
+
+        expect:
+        engine.eval(input, evalContext).get() == expResult
+
+        where:
+        input             || expResult
```

Comment:
! is always at the beginning.
Please add some tests where negation is inside the expression in various places to test that operator priority works fine